canopy-cover-enhancement/terra_ganEnhancement.py

```diff
@@ -0,0 +1,262 @@
+        im2[0: h - scales[s], :] = im[scales[s]:h, :]
+        dif = im * (im1 - im2)
+        FM[s] = np.mean(dif)
+    NRMAC = np.mean(FM)
+    return NRMAC
+
+
+# check how many percent of pix close to 255 or 0
+def check_saturation(img):
+    grayImg = cv2.cvtColor(img, cv2.COLOR_BGR2GRAY)
+
+    m1 = grayImg > SATURATE_THRESHOLD
+    m2 = grayImg < 20  # 20 is a threshold to classify low pixel value
+
+    over_rate = float(np.sum(m1)) / float(grayImg.size)
+    low_rate = float(np.sum(m2)) / float(grayImg.size)
+
+    return over_rate, low_rate
+
+
+# gen average pixel value from grayscale image
+def check_brightness(img):
+    grayImg = cv2.cvtColor(img, cv2.COLOR_BGR2GRAY)
+
+    aveValue = np.average(grayImg)
+
+    return aveValue
+
+
+# abandon low quality images, mask enhanced
+def gen_cc_enhanced(input_path, kernelSize=3):
+    img = cv2.imread(input_path)
+
+    # calculate image scores
+    over_rate, low_rate = check_saturation(img)
+
+    aveValue = check_brightness(img)
+
+    quality_score = getImageQuality(input_path)
+
+    # if low score, return None
+    # low_rate is percentage of low value pixels(lower than 20) in the grayscale image, if low_rate > 0.1, return
+    # aveValue is average pixel value of grayscale image, if aveValue lower than 30 or higher than 195, return
+    # quality_score is a score from Multiscale Autocorrelation (MAC), if quality_score lower than 13, return
+    if low_rate > 0.1 or aveValue < 30 or aveValue > 195 or quality_score < 13:
+        return None, None, None
+
+    # saturated image process
+    # over_rate is percentage of high value pixels(higher than SATUTATE_THRESHOLD) in the grayscale image, if over_rate > 0.15, try to fix it use gen_saturated_mask()
+    if over_rate > 0.15:
+        binMask = gen_saturated_mask(img, kernelSize)
+    else:  # nomal image process
+        binMask = gen_mask(img, kernelSize)
+
+    c = np.count_nonzero(binMask)
+    ratio = c / float(binMask.size)
+
+    rgbMask = gen_rgb_mask(img, binMask)
+
+    return ratio, binMask, rgbMask
+
+
+class ganEnhancementExtractor(TerrarefExtractor):
+
+    # TODO should we ignore winter wheat seasons?
+
+    def __init__(self):
+        super(ganEnhancementExtractor, self).__init__()
+
+        add_local_arguments(self.parser)
+
+        # assign local arguments
+        self.SATURATE_THRESHOLD = self.args.saturate_threshold
+        self.MAX_PIXEL_VAL = self.args.max_pixel_val
+        self.SMALL_AREA_THRESHOLD = self.args.small_area_threshold
+
+        # parse command line and load default logging configuration
+        self.setup(sensor='ganEnhancement')
+
+    def check_message(self, connector, host, secret_key, resource, parameters):
+        # TODO not sure what rules should be checked here, returning true
+        return CheckMessage.download
+
+    def process_message(self, connector, host, secret_key, resource, parameters):
+        self.start_message(resource)
+        for fname in resource['local_paths']:
```

Comment:
we dont just want to do every file because they might not be relevant TIF files.

https://github.com/terraref/extractors-stereo-rgb/blob/master/bin2tif/terra_bin2tif.py#L68

Use a conditional like this to find _left.tif and _right.tif and metadata.